### PR TITLE
Add granular data to the HTML report graphs

### LINKED ISSUE #410: Display more granular data on graphs
At the moment we only display total (aggregated) numbers. It would be great to also include more granular data (requests per request type, response time per request type, executed tasks per task set/task type, ....).

### PR DESCRIPTION
**Pull request by slashrsm**
Closes #410.

I used the default color scheme for now. I don't particularly like it, but I didn't have any better idea either.

Screenshot 2022-02-10 at 23 34 32
Screenshot 2022-02-10 at 23 34 18

### CODE REVIEW

**Comment by jeremyandrews** [comment]
How difficult would it be to add a flag to disable this verbose data? It creates much larger report files, and perhaps some would prefer the simpler and smaller graphs that we've generated up to this PR? I'd leave granular data enabled by default, but it could be useful to have an option like `--no-granular-data` or something, especially useful on larger/longer load tests?

File: src/graph.rs [hunk 1]
```diff
@@ -331,7 +382,7 @@ impl<'a, T: Serialize> Graph<'a, T> {
                     var myChart = echarts.init(chartDom);
 
                     myChart.setOption({{
-                        color: ['#2c664f'],
+                        color: ['#2c664f', '#5470c6', '#91cc75', '#fac858', '#ee6666', '#73c0de', '#3ba272', '#fc8452', '#9a60b4', '#ea7ccc'],
```
[the remark this answers is not in the document]
Comment (slashrsm) [reply src/graph.rs:385]: Correct. We could add more, but at some point they will start repeating anyway. 

Maybe there is a way to set these programmatically? A follow-up maybe?

File: src/graph.rs [hunk 2]
```diff
@@ -72,122 +72,122 @@ impl GraphData {
     }
 
     /// Record requests per second metric.
-    pub(crate) fn record_requests_per_second(&mut self, second: usize) {
-        expand_per_second_metric_array(&mut self.requests_per_second, second, 0);
-        self.requests_per_second[second] += 1;
+    pub(crate) fn record_requests_per_second(&mut self, key: String, second: usize) {
```
[the remark this answers is not in the document]
Comment (slashrsm) [reply src/graph.rs]: Done.

File: src/graph.rs [hunk 3]
```diff
@@ -72,122 +72,122 @@ impl GraphData {
...
+            second,
+            data.get(second)
         );
     }
 
     /// Record errors per second metric.
-    pub(crate) fn record_errors_per_second(&mut self, second: usize) {
-        expand_per_second_metric_array(&mut self.errors_per_second, second, 0);
-        self.errors_per_second[second] += 1;
+    pub(crate) fn record_errors_per_second(&mut self, key: String, second: usize) {
```
[the remark this answers is not in the document]
Comment (slashrsm) [reply src/graph.rs]: Done.

File: src/graph.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (slashrsm) [reply src/graph.rs]: And an excellent suggestion!

**Comment by nicompte** [comment]
Hello there, just a suggestion after testing your branch, do you think it would be possible to keep the same color for the same request, between two runs?

File: src/graph.rs [hunk 4]
```diff
@@ -322,6 +314,65 @@ impl<'a, T: Serialize> Graph<'a, T> {
...
+            for (_, single_data) in self.data.iter() {
+                total_values.add(single_data);
+            }
+
+            // We will have multiple lines. We need to prepare the legend section on the graph
+            // and create data series for all of them.
+            let mut legend = vec!["Total"];
+
+            let mut other_values = String::new();
+            for (title, sub_data) in self.data.iter() {
```
[the remark this answers is not in the document]
Comment (slashrsm) [reply src/graph.rs]: Done.

File: src/graph.rs [hunk 5]
```diff
@@ -383,27 +436,205 @@ impl<'a, T: Serialize> Graph<'a, T> {
...
+        };
+    }
+}
+
+/// Defines a single value in a TimeSeries.
+pub trait TimeSeriesValue<T, U> {
+    /// Initial ("zero") value.
+    fn initial_value() -> T;
+    /// Adds the given value to the current value.
+    fn add(&mut self, value: U);
```
[the remark this answers is not in the document]
Comment (slashrsm) [reply src/graph.rs]: Done.

File: src/graph.rs [hunk 6]
```diff
@@ -383,27 +436,205 @@ impl<'a, T: Serialize> Graph<'a, T> {
...
+}
+
+/// Data structure to represent time series data.
+#[derive(Debug, Clone, PartialEq)]
+struct TimeSeries<T: TimeSeriesValue<T, U>, U> {
+    /// Time series data.
+    ///
+    /// Each element of the vector represents value for one second in the time series.
+    data: Vec<T>,
+    /// Phantom data indicates to the compiler that the "U" generic data type has zero size.
```
[the remark this answers is not in the document]
Comment (slashrsm) [reply src/graph.rs:540]: I don't think so. It appears to be zero-size:

> PhantomData consumes no space, but simulates a field of the given type for the purpose of static analysis.

https://doc.rust-lang.org/nomicon/phantom-data.html

Fun fact - It appears that it is used both in `Iter` and `Vec<T>`.

**Comment by nicompte** [comment]
> Good point. I added ordering to the lines/legend based on the numbers behind them. I suspect that same order also means same colors so this might solve your suggestion too. @nicompte Do you mind testing a bit?

Works great thanks!

File: src/graph.rs [hunk 7]
```diff
@@ -72,122 +72,122 @@ impl GraphData {
...
-            &response_times,
+            self.average_response_time_per_second.clone(),
         )
     }
 
     /// Generate active tasks graph.
-    pub(crate) fn get_tasks_per_second_graph(&self) -> Graph<usize> {
-        self.create_graph_from_data("graph-tps", "Tasks #", &self.tasks_per_second)
+    pub(crate) fn get_tasks_per_second_graph(&self) -> Graph<usize, usize> {
+        self.create_graph_from_single_data("graph-tps", "Tasks #", self.tasks_per_second.clone())
```
[the remark this answers is not in the document]
Comment (jeremyandrews) [reply src/graph.rs]: per-task?

File: src/config.rs [hunk 8]
```diff
@@ -59,6 +59,7 @@ const DEFAULT_PORT: &str = "5115";
 /// --no-error-summary         Doesn't display an error summary
 /// --no-print-metrics         Doesn't display metrics at end of load test
 /// --report-file NAME         Create an html-formatted report
+/// --no-granular-data         Disables granular data in HTML report graphs
```
Comment (jeremyandrews) [inline src/config.rs:62]: Would `--no-granular-report` make it more clear that this specifically impacts `--report-file` and no other data?

File: src/graph.rs [hunk 9]
```diff
@@ -20,18 +24,16 @@ pub(crate) struct GraphData {
...
-    /// Maintains average response time per second. Each element of the vector represents one second.
-    average_response_time_per_second: Vec<MovingAverage>,
-    /// Number of tasks at the end of each second of the test. Each element of the vector
-    /// represents one second.
-    tasks_per_second: Vec<usize>,
-    /// Number of users at the end of each second of the test. Each element of the vector
-    /// represents one second.
-    users_per_second: Vec<usize>,
+    /// Counts requests per second for each request type.
+    requests_per_second: HashMap<String, TimeSeries<u32, u32>>,
```
Comment (jeremyandrews) [inline src/graph.rs:28]: (Followup is fine) it seems worthwhile to create a custom Type for `HashMap<String, TimeSeries<u32, u32>>`, something like `ItemsPerSecond` or `PerSecondCounter` or ...? I see it throughout the new code, and we'd document it thoroughly where the type is created.

**Review by jeremyandrews — APPROVED** [review]
This is great, thanks! I'm doing more testing, but there's nothing blocking a merge now. I think the configuration option change is worthwhile to be more clear

**Pull request merged**